File: Solutions/java/y2015/Solution01.java
package Solutions.java.y2015;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import Solutions.java.Solution;

public class Solution01 implements Solution {

    @Override
    public void Part1(String inputFile) {
        Path filename = Path.of(inputFile);
        var result = 0;

        try {
            var data = Files.readString(filename);
            for (int i =0; i < data.length(); ++i) {
                result += data.charAt(i) == '(' ? 1 : -1;
            }
        }
        catch(IOException exception) {

        }

        System.out.printf("Santa is on the %dth floor\n", result);
    }

    @Override
    public void Part2(String inputFile) {
        Path filename = Path.of(inputFile);
        var result = 0;
        var tryCounter = 0;

        try {
            var data = Files.readString(filename);
            for (int i =0; i < data.length(); ++i) {
                result += data.charAt(i) == '(' ? 1 : -1;
                tryCounter++;

                if(result < 0) {
                    break;
                }
            }
        }
        catch(IOException exception) {

        }

        System.out.printf("Santa found the basement after %d tries", tryCounter);
    }
    
}

File: Solutions/kotlin/y2015/solution01.kt
package Solutions.kotlin.y2015

import Solutions.kotlin.Solution
import java.io.InputStream
import java.io.File
import java.io.BufferedReader

class Solution01 : Solution
{
    override fun Part1(inputFile: String) {
        val data = File(inputFile).readText()
        var result = data.count { it == '(' } - data.count { it == ')' }

        print("Santa is on the %dth floor\n".format(result));
    }
    
    override fun Part2(inputFile: String) {
        var result = 0
        var tryCounter = 0

        File(inputFile).useLines() { 
            lines -> lines.forEach { 
                for (c in it) {
                    result += if(c == '(') 1 else -1
                    tryCounter++

                    if(result < 0) {
                        break
                    }
                } 
            } 
        }

        print("Santa found the basement after %d tries\n".format(tryCounter));
    }
}